File: C++/Data structure/graph/cycle_in_directed_graph.cpp
#include<iostream>
#include<bits/stdc++.h>
using namespace std;

//Time complexity O(V+E)
//Space complexity O(V)

int isCycleUtil(vector<int> adjList[],vector<bool> visited,int curr){
    if(visited[curr]==true)
       return true;
    visited[curr]=true;
       for(int j=0;j < adjList[curr].size();j++){
           if(isCycleUtil(adjList,visited,adjList[curr][j]))
                 return true;
       }
       return false;
}

int isCycle(vector<int> adjList[],int V){
    vector<bool> visited(V,false);
    for(int i=0;i < V;i++){
        visited[i] = true;
       for(int j=0;j < adjList[i].size();j++){
           if(isCycleUtil(adjList,visited,adjList[i][j]))
               return true;
       }
       visited[i]= false;
    }
    return false;
}



int main()
{
    int V,E,src,dst;
    cout << "enter number of vertices:";
    cin >> V;
    cout << "enter number of Edges:";
    cin >> E;
    
    vector<int> adjList[V];
    for(int i =0;i<E;i++)
    {
        cout << "enter edge src dst:";
        cin >> src >> dst;
        // directed graph
        adjList[src].push_back(dst);
    }
    cout << isCycle(adjList,V);
}

File: Python/Data_Structure/graph/cycle_in_directed_graph.py
from collections import defaultdict

class Graph:
    def __init__(self,V):
        self.vertices = V
        self.node = defaultdict(list)

    def add_edges_directed_graph(self,src,dst):
        self.node[src].append(dst)

    def printList(self):
        print(self.node)

    def isCycle(self):
        visited = [False]*self.vertices
        for i in range(self.vertices):
            visited[i] = True
            for j in range(len(self.node[i])):
                if(self.isCycleUtil(self.node[i][j],visited)):
                    return True
            visited[i] = False
        return False
        
    def isCycleUtil(self,curr,visited):
        if visited[curr] == True:
            return True
        visited[curr] = True
        for i in range(len(self.node[curr])):
            if(self.isCycleUtil(self.node[curr][i],visited.copy())):
                return True
        return False


        

if __name__ == "__main__":
    print()
    V = int(input("Enter number of vertices:"))
    E = int(input("Enter number of Edges:"))
    graph = Graph(V)
    print("Enter Edges:-")
    for i in range(E):
        src,dst = map(int,input().split())
        graph.add_edges_directed_graph(src,dst)
    print(graph.printList())
    print(graph.isCycle())

